File: java/beer-song/src/main/java/BeerSong.java
public class BeerSong{

    private String verse;
    private String penultimateVerse;
    private String finalVerse;

    public BeerSong() {
        this.verse = "%d bottles of beer on the wall, %d bottles of beer.\n" +
                        "Take one down and pass it around, %d %s of beer on the wall.\n\n";
        this.penultimateVerse = "1 bottle of beer on the wall, 1 bottle of beer.\n" +
                        "Take it down and pass it around, no more bottles of beer on the wall.\n\n";
        this.finalVerse = "No more bottles of beer on the wall, no more bottles of beer.\n" + 
                        "Go to the store and buy some more, 99 bottles of beer on the wall.\n\n";
    }

    public String sing(int bottle, int numberOfVerses) {

        StringBuilder sb = new StringBuilder();

        for(int i = bottle; i > bottle - numberOfVerses; i--) {
            if (i == 1) {
                sb.append(penultimateVerse);
            } else if (i == 0) {
                sb.append(finalVerse);
            } else {
                String actualVerse = String.format(verse, i, i, 
                                        i - 1, (i > 2 ? "bottles" : "bottle"));
                sb.append(actualVerse);
            }
        }

        return sb.toString();
    }

    public String singSong() {
        return sing(99, 100);
    }
}

File: elixir/beer-song/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(0),
    do: """
    No more bottles of beer on the wall, no more bottles of beer.
    Go to the store and buy some more, 99 bottles of beer on the wall.
    """

  def verse(1),
    do: """
    1 bottle of beer on the wall, 1 bottle of beer.
    Take it down and pass it around, no more bottles of beer on the wall.
    """

  def verse(2),
    do: """
    2 bottles of beer on the wall, 2 bottles of beer.
    Take one down and pass it around, 1 bottle of beer on the wall.
    """

  def verse(number),
    do: """
    #{number} bottles of beer on the wall, #{number} bottles of beer.
    Take one down and pass it around, #{number - 1} bottles of beer on the wall.
    """

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0) do
    Enum.reduce(range, "", fn verse_number, acc -> acc <> "\n" <> verse(verse_number) end)
    |> String.trim_leading("\n")
  end
end
